include/libopencm3/stm32/common/adc_common_v2.h

```diff
@@ -36,138 +36,179 @@ specific memorymap.h header before including this header file.*/
+/** ADSTART: ADC start of regular conversion */
 #define ADC_CR_ADSTART		(1 << 2)
-/* ADDIS: ADC disable command */
+/** ADDIS: ADC disable command */
 #define ADC_CR_ADDIS		(1 << 1)
-/* ADEN: ADC enable control */
+/** ADEN: ADC enable control */
 #define ADC_CR_ADEN		(1 << 0)
 
+/**@}*/
+
 /* ADC_CFGR1 Values -----------------------------------------------------------*/
+/** @defgroup adc_cfgr1 CFGR1 ADC configuration register 1
+@{*/
 
-/* AWD1CH[4:0]: Analog watchdog 1 channel selection */
 #define ADC_CFGR1_AWD1CH_SHIFT		26
 #define ADC_CFGR1_AWD1CH			(0x1F << ADC_CFGR1_AWD1CH_SHIFT)
+/** AWD1CH: Analog watchdog 1 channel selection */
 #define ADC_CFGR1_AWD1CH_VAL(x)		((x) << ADC_CFGR1_AWD1CH_SHIFT)
 
-/* AWD1EN: Analog watchdog 1 enable on regular channels */
+/** AWD1EN: Analog watchdog 1 enable on regular channels */
 #define ADC_CFGR1_AWD1EN		(1 << 23)
-
-/* AWD1SGL: Enable the watchdog 1 on a single channel or on all channels */
+/** AWD1SGL: Enable the watchdog 1 on a single channel or on all channels */
 #define ADC_CFGR1_AWD1SGL		(1 << 22)
-
-/* DISCEN: Discontinuous mode for regular channels */
+/** DISCEN: Discontinuous mode for regular channels */
 #define ADC_CFGR1_DISCEN		(1 << 16)
-
-/* AUTDLY: Delayed conversion mode */
+/** AUTDLY: Delayed conversion mode */
 #define ADC_CFGR1_AUTDLY		(1 << 14)
-
-/* CONT: Single / continuous conversion mode for regular conversions */
+/** CONT: Single / continuous conversion mode for regular conversions */
 #define ADC_CFGR1_CONT		(1 << 13)
-
-/* OVRMOD: Overrun Mode */
+/** OVRMOD: Overrun Mode */
 #define ADC_CFGR1_OVRMOD		(1 << 12)
 
-/*
- * EXTEN[1:0]: External trigger enable and polarity selection for regular
- * channels
- */
+#define ADC_CFGR1_EXTEN_MASK		(0x3 << 10)
+/** @defgroup adc_cfgr1_exten EXTEN: External trigger enable and polarity selection for regular channels
+@{*/
 #define ADC_CFGR1_EXTEN_DISABLED		(0x0 << 10)
 #define ADC_CFGR1_EXTEN_RISING_EDGE	(0x1 << 10)
 #define ADC_CFGR1_EXTEN_FALLING_EDGE	(0x2 << 10)
 #define ADC_CFGR1_EXTEN_BOTH_EDGES	(0x3 << 10)
+/**@}*/
 
-#define ADC_CFGR1_EXTEN_MASK		(0x3 << 10)
-
-/* ALIGN: Data alignment */
+/** ALIGN: Data alignment */
 #define ADC_CFGR1_ALIGN		(1 << 5)
 
-/* RES[1:0]: Data resolution */
+#define ADC_CFGR1_RES_MASK	(0x3 << 3)
+/** @defgroup adc_cfgr1_res RES: Data resolution
+@{*/
 #define ADC_CFGR1_RES_12_BIT	(0x0 << 3)
 #define ADC_CFGR1_RES_10_BIT	(0x1 << 3)
 #define ADC_CFGR1_RES_8_BIT	(0x2 << 3)
 #define ADC_CFGR1_RES_6_BIT	(0x3 << 3)
-#define ADC_CFGR1_RES_MASK	(0x3 << 3)
+/**@}*/
 
-/* DMACFG: Direct memory access configuration */
+/** DMACFG: Direct memory access configuration */
 #define ADC_CFGR1_DMACFG		(1 << 1)
 
-/* DMAEN: Direct memory access enable */
+/** DMAEN: Direct memory access enable */
 #define ADC_CFGR1_DMAEN		(1 << 0)
 
+/**@}*/
+
+/* ADC_SMPR Values -----------------------------------------------------------*/
+/** @defgroup adc_smpr SMPR ADC sample time register
+@{*/
```

Comment:
you open a group here without closing it?  and really, if you're not putting things in, shouldn't we just drop them?

Reply:
oops, fixed - pushed too quick, updated pr with closed group.

My idea here was: we define a register, let's define also the maching doxygen group too, and avoid doing it in each device's adc.h when defining register bit fields. ppl can then just @addtogroup, like i did in g0 adc. But it can look a bit weird, indeed, when no single bit it shared between chips. Up to you, i can drop that.